### ISSUE TITLE
Support changelogs from bitbucket-server

### ISSUE THREAD

**Issue by k2snowman69** [renovatebot/renovate#3903]
**What would you like Renovate to be able to do?**
Similar to how renovate pulls from github.com to retrieve existing changelogs, it would be great if it did similar for on-premise versions of bitbucket-server

**Describe the solution you'd like**
Input - A bitbucket-server repository with a changelog.md at the root
Output - When a PR is created, it uses that changelog to improve the PR body created

**Describe alternatives you've considered**
I thought about other locations where we maybe able to get the changelog from such as unpkg or artifactory if it was part of the npm package but I feel that it adds more configuration changes and dependencies than less

**Additional context**
N/A

**Comment by k2snowman69** [comment]
Haven't done much investigation into this but I saw the one for gitlab and maven so I figured we should open one for bitbucket-server as well

**Comment by rarkins** [comment]
No, it would be a separate task because the APIs are quite different

**Comment by Zenedith** [comment]
Should [this](https://github.com/renovatebot/renovate/pull/7000) help?

**Comment by viceice** [comment]
@Zenedith no, as that is only for shared presets

before we can start working on this, the changelog retrival code must be refctored, so we can add more platofrms more easy

**Comment by jaysiee** [comment]
Any chance someone less experienced could contribute to the refactoring? If there is enough interest, I'd be happy to help.

**Comment by k2snowman69** [comment]
Since no one else chimed in yet... I'd say it depends on what you mean by "less experienced" but as the old adage goes, "you never know until you try".

From my personal experience creating my first PR a few months back, I'd say the biggest hurdle is just getting this project running in a way that you can debug it in a way that's comfortable for you. Everyone has their own coding style so getting this project to run with your personal style along with in a way that doesn't screw up a project in actual use.

After that though, I found it pretty easy to probe around the code base and figure out what's going on where.

**Comment by jaysiee** [comment]
Thanks for your feedback. I'm fairly opimistic that I'll get the project up and running. The part I that threw me off was the required refactoring and whether this is realistically something that can be done by someone new to the project (not knowing what the maintainers have in mind). 

I've had a quick look and currently the release-notes part is branching at multiple points based on whether it should work for github or gitlab. Since it is not intended to simply add another conditional branch for bitbucket-server, maybe we could sketch out the code structure that is required for a more general approach to fetching changelogs and then someone less involved could give it a shot. 

Alternatively, I could also just go ahead and work on a more general approach to fetching changelogs myself and create a draft PR (in that case I'd still be happy for any hints regarding potential stumbling blocks).

**Comment by rarkins** [comment]
@viceice do you have any objection if @jaysiee adds in BB prior to refactoring?

**Comment by viceice** [comment]
Without refactoring the code would get more complicated as @jaysiee already mentioned.

 Bu i also know there is currently nobody who will and can do the required refactoring. So maybe @jaysiee  can go ahead.
we can split his pr later if required.
